From 2f3161ba7caad9494cff694e290d7b2d376e3271 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Wed, 9 Dec 2015 18:35:08 +0000
Subject: [PATCH 1/2] scripts/functions: Use blkid to resolve LABEL= and UUID=
 block device IDs

This is a preparatory patch for fixing #801154.

Ben.
---
 scripts/functions | 32 +++++---------------------------
 1 file changed, 5 insertions(+), 27 deletions(-)

diff --git a/scripts/functions b/scripts/functions
index 8c1bb1f..d724f74 100644
--- a/scripts/functions
+++ b/scripts/functions
@@ -294,35 +294,13 @@ read_fstab_entry() {
 resolve_device() {
 	DEV="$1"
 
-	case $DEV in
-	LABEL=*)
-		DEV="${DEV#LABEL=}"
-
-		# support any / in LABEL= path (escape to \x2f)
-		case "${DEV}" in
-		*/*)
-		if command -v sed >/dev/null 2>&1; then
-			DEV="$(echo ${DEV} | sed 's,/,\\x2f,g')"
+	case "$DEV" in
+	LABEL=* | UUID=*)
+		if command -v blkid >/dev/null 2>&1; then
+			DEV="$(blkid -l -t "$DEV" -o device)"
 		else
-			if [ "${DEV}" != "${DEV#/}" ]; then
-				DEV="\x2f${DEV#/}"
-			fi
-			if [ "${DEV}" != "${DEV%/}" ]; then
-				DEV="${DEV%/}\x2f"
-			fi
-			IFS='/'
-			newroot=
-			for s in $DEV; do
-				newroot="${newroot:+${newroot}\\x2f}${s}"
-			done
-			unset IFS
-			DEV="${newroot}"
+			log_warning_msg "blkid not present, so cannot resolve $DEV"
 		fi
-		esac
-		DEV="/dev/disk/by-label/${DEV}"
-		;;
-	UUID=*)
-		DEV="/dev/disk/by-uuid/${DEV#UUID=}"
 		;;
 	esac
 	# Only canonicalise if a valid file, in case $DEV isn't a filename
-- 
2.6.2

